### PR TITLE
Optimize InsertTensors into TensorViews with Offsets.

### PR DESCRIPTION
**Pull request by jfix71**
The main aim here is optimizing the pattern seen by Concats IRGen'd into InsertTensor. If all inserts are into the first dimension, then we can replace the InsertTensor with a TensorView with the same offset as the InsertTensor. This is because the underlying memory is contiguous for the InsertTensor/TensorView.

This improves ptb execution time on the CPU backend from 7.1 -> 6.2 seconds, which has a big ConcatNode.

For now it only applies to insert tensors with an alloc as the input; I believe this is most common/important.

### CODE REVIEW

File: lib/Optimizer/IROptimizer.cpp [hunk 1]
```diff
@@ -1123,6 +1142,106 @@ static void eliminateDeadStores(IRFunction &M) {
...
+    auto *I = *it;
+    // Look for compatible InsertTensors.
+    if (auto *ITI = dyn_cast<InsertTensorInst>(I)) {
+      // TVI with offsets currently only works for this optimization if the
+      // insert is only into the first dimension. This is because the tensor
+      // memory must be contiguous.
+      if (!isOnlyOffsetInFirstDim(ITI->getOffsets())) {
+        continue;
+      }
+      if (auto *insertSourceAAI =
```
Comment (opti-mix) [inline lib/Optimizer/IROptimizer.cpp]: Use `if (!cond) continue;` as the early exit to reduce the indentation.

File: lib/Optimizer/IROptimizer.cpp [hunk 2]
```diff
@@ -1123,6 +1142,106 @@ static void eliminateDeadStores(IRFunction &M) {
...
+              dyn_cast<AllocActivationInst>(ITI->getSrc())) {
+        // Assume 3 users of the source AAI (ITI, the original instruction that
+        // writes into it, and a dealloc). This is the pattern generated by an
+        // IRGen'd ConcatNode.
+        if (insertSourceAAI->getNumUsers() == 3) {
```
Comment (opti-mix) [inline lib/Optimizer/IROptimizer.cpp]: It is probably not enough to check only the number of uses. You may need to check exactly what the comment above states, i.e. that there is an alloc, the write and the dealloc.

File: lib/Optimizer/IROptimizer.cpp [hunk 3]
```diff
@@ -1123,6 +1142,106 @@ static void eliminateDeadStores(IRFunction &M) {
...
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+/// \returns true if the dimensions of \p sourceDims, other than the first
+/// dimension, all equal the dimensions in \p destDims.
+bool allButFirstDimsEqual(llvm::ArrayRef<size_t> sourceDims,
```
Comment (opti-mix) [inline lib/Optimizer/IROptimizer.cpp]: Should it be static?

File: lib/Optimizer/IROptimizer.cpp [hunk 4]
```diff
@@ -1123,6 +1142,106 @@ static void eliminateDeadStores(IRFunction &M) {
   eraseInstructions(M, erasedInstructions);
 }
 
+/// \returns true if the first dimension in \p offsets has a non-zero value.
+bool isOnlyOffsetInFirstDim(llvm::ArrayRef<size_t> offsets) {
```
Comment (opti-mix) [inline lib/Optimizer/IROptimizer.cpp]: Should it be static?

File: lib/Optimizer/IROptimizer.cpp [hunk 5]
```diff
@@ -1246,8 +1366,12 @@ void performPeepholeOptimizations(IRFunction &M) {
 
     // Remove useless copies.
     if (auto *CI = dyn_cast<CopyInst>(I)) {
-      if (getOrigin(CI->getSrc()) == getOrigin(CI->getDest()))
+      auto srcOrigin = getOrigin(CI->getSrc());
+      auto destOrigin = getOrigin(CI->getDest());
+      if ((srcOrigin == destOrigin) &&
+          (srcOrigin->getType()->size() == destOrigin->getType()->size())) {
```
Comment (opti-mix) [inline lib/Optimizer/IROptimizer.cpp]: If srcOrigin == destOrigin, then `srcOrigin->getType()->size() == destOrigin->getType()->size()` is always true, or? Why do we need the check it? Or may be you meant a different condition?

File: include/glow/IR/IRBuilder.h [hunk 6]
```diff
@@ -58,6 +58,11 @@ class IRBuilder {
                                    llvm::ArrayRef<size_t> dims, Value *src,
                                    llvm::StringRef name);
 
+  TensorViewInst *createTensorView(ElemKind elemKind,
```
Comment (opti-mix) [inline include/glow/IR/IRBuilder.h]: Do we need both forms of `createTensorView`? May be the new form with offsets can take offset as an optional parameter with default value (e.g. an empty set of offsets)?

File: lib/Optimizer/IROptimizer.cpp (hunk 2 above)
Comment (jfix71) [reply lib/Optimizer/IROptimizer.cpp]: Above I already have checked if the insert's source is an `AllocActivationInst` which accounts for one use, and then assuming well formedness there must be a second use that is `DeallocActivationInst`. So this means the third use must be the insert here. Do you think explicitly verifying `DeallocActivationInst` is necessary?

**Comment by jfix71** [comment]
All changes have been made.

Not including compilation time, on the CPU backend ptb goes from 5.1 -> 4.2 seconds.

**Pull request merged**